File: dotnet/Allmark/Block/BlockQuoteRule.cs
namespace Allmark.Block;

using Allmark.Parse;
using Allmark.Types;

/// <summary>
/// A block quote is a sequence of lines that is considered to be a single unit.
/// </summary>
public static class BlockQuoteRule
{
	public static BlockRule Create()
	{
		return new BlockRule
		{
			Name = "block_quote",
			TestStart = TestStart,
			TestContinue = TestContinue,
			CloseNode = Close
		};
	}

	private static bool HasMarkup(char c, BlockParserState state)
	{
		return state.Indent <= 3 && c == '>';
	}

	private static bool TestStart(BlockParserState state, MarkdownNode parent)
	{
		MarkdownNode? closedNode = null;

		if (parent.AcceptsContent)
		{
			return false;
		}

		char c = Utils.GetChar(state.Src, state.I);
		if (HasMarkup(c, state))
		{
			if (parent.Type == "paragraph")
			{
				closedNode = state.OpenNodes.Pop();
				parent = state.OpenNodes.Peek();
			}

			if (closedNode != null)
			{
				Utils.CloseNode(state, closedNode);
			}

			int quoteIndent = state.Indent + 1;

			var quote = Utils.NewNode("block_quote", true, state.I, state.Line, 1, c.ToString(), quoteIndent, new List<MarkdownNode>());

			parent.Children!.Add(quote);
			state.OpenNodes.Push(quote);

			Utils.MovePastMarker(1, state);

			state.HasBlankLine = false;
			ParseBlock.Execute(state, quote);

			return true;
		}

		return false;
	}

	private static bool TestContinue(BlockParserState state, MarkdownNode node)
	{
		char c = Utils.GetChar(state.Src, state.I);
		if (HasMarkup(c, state))
		{
			Utils.MovePastMarker(1, state);
			return true;
		}

		if (state.HasBlankLine)
		{
			return false;
		}

		var openNode = state.OpenNodes.Peek();
		if (openNode.Type == "paragraph")
		{
			state.MaybeContinue = true;
			node.MaybeContinuing = true;
			return true;
		}

		return false;
	}

	private static void Close(BlockParserState state, MarkdownNode node)
	{
		if (state.HasBlankLine && node.Children != null && node.Children.Count > 0)
		{
			node.Children[^1].BlankAfter = true;
			state.HasBlankLine = false;
		}
	}
}

File: web/src/block/blockQuoteRule.ts
import parseBlock from "../parse/parseBlock";
import type BlockParserState from "../types/BlockParserState";
import type BlockRule from "../types/BlockRule";
import type MarkdownNode from "../types/MarkdownNode";
import { ANGLE_RIGHT_CODE } from "../utils/charCodes";
import closeNode from "../utils/closeNode";
import movePastMarker from "../utils/movePastMarker";
import newBlock from "../utils/newBlock";
import { appendChild, getLastChild, hasChildren } from "../utils/nodeUtils";

const rule: BlockRule = {
	name: "block_quote",
	testStart,
	testContinue,
	closeNode: close,
};
export default rule;

/**
 * "A block quote marker consists of 0-3 spaces of initial indent, plus (a) the
 * character > together with a following space, or (b) a single character > not
 * followed by a space.
 *
 * The following rules define block quotes:
 *
 * 1. Basic case. If a string of lines Ls constitute a sequence of blocks Bs,
 *    then the result of prepending a block quote marker to the beginning of
 *    each line in Ls is a block quote containing Bs.
 *
 * 2. Laziness. If a string of lines Ls constitute a block quote with contents
 *    Bs, then the result of deleting the initial block quote marker from one or
 *    more lines in which the next non-whitespace character after the block
 *    quote marker is paragraph continuation text is a block quote with Bs as
 *    its content. Paragraph continuation text is text that will be parsed as
 *    part of the content of a paragraph, but does not occur at the beginning of
 *    the paragraph.
 *
 * 3. Consecutiveness. A document cannot contain two block quotes in a row
 *    unless there is a blank line between them.
 *
 * Nothing else counts as a block quote."
 */

function hasMarkup(state: BlockParserState) {
	return state.indent <= 3 && state.src.charCodeAt(state.i) === ANGLE_RIGHT_CODE;
}

function testStart(state: BlockParserState, parent: MarkdownNode) {
	let closedNode: MarkdownNode | undefined;

	if (parent.acceptsContent) {
		return false;
	}

	if (hasMarkup(state)) {
		if (parent.type === "paragraph") {
			closedNode = state.openNodes.pop();
			parent = state.openNodes.at(-1)!;
		}

		if (closedNode !== undefined) {
			closeNode(state, closedNode);
		}

		let quoteIndent = state.indent + 1;

		let quote = newBlock("block_quote", state.i, state.line, ">", quoteIndent);

		appendChild(parent, quote);
		state.openNodes.push(quote);

		movePastMarker(1, state);

		state.hasBlankLine = false;
		parseBlock(state, quote);

		return true;
	}

	return false;
}

function testContinue(state: BlockParserState, node: MarkdownNode) {
	if (hasMarkup(state)) {
		movePastMarker(1, state);
		return true;
	}

	if (state.hasBlankLine) {
		return false;
	}

	let openNode = state.openNodes.at(-1)!;
	if (openNode.type === "paragraph") {
		// We won't know until we try more things
		state.maybeContinue = true;
		node.maybeContinuing = true;
		return true;
	}

	return false;
}

function close(state: BlockParserState, node: MarkdownNode) {
	// Swallow blank lines
	if (state.hasBlankLine && hasChildren(node)) {
		getLastChild(node)!.blankAfter = true;
		state.hasBlankLine = false;
	}
}
